posts/swift-for-cxx-practitioners-type-erasure/index.html

```diff
@@ -185,7 +185,7 @@
-</code></pre><p>Each <code>some</code> type introduces an unnamed generic parameter whose constraints are listed after the <code>some</code>. Swift <code>some</code> types are also called <em>opaque</em> types, because they hide the name of the underlying type: an unnamed generic parameter can't be named (duh), so the actual collection type based into <code>operateOnAny(strings:)</code> is hidden from the function body. Opaque types are also useful in the return type of a function to hide the specific return type from the caller.</p><h2>Implementation hiding with opaque types</h2><p><code>any</code> types effectively hide the underlying type from clients, allowing it to change dynamically at run time. These are two different things: the first is about hiding implementation details behind an abstraction barrier (e.g., we know that we have a <code>Collection</code> of <code>String</code>s, but not the specific type of the collection itself) and the other is about allowing the underlying representation to change dynamically (e.g., we can choose <code>Set&lt;String&gt;</code> or <code>[String]</code> depending on what's best for the problem at hand).</p><p>Swift has a notion of <em>opaque</em> types that let you hide the implementation type behind an abstraction barrier without allowing it to change dynamically at runtime. This gives you, as the implementer, the freedom to limit your API surface area (by not exposing some types) and evolve your implementations over time, without breaking clients. For example, let's consider implementing a generic function <code>uniqued</code> on a <code>Collection</code> that produces a new collection with duplicates removed. We could have such a function return a <code>Set</code>:</p><pre><code><span class="keyword">extension</span> <span class="type">Collection</span> <span class="keyword">where</span> <span class="type">Value</span>: <span class="type">Hashable</span> {
```

Comment:
I corrected the word 'some' to 'specific' so it cannot be confused with the opaque type keyword `some`.

Reply:
Good idea!